File: data/leetCode Solutions/Q3_lengthOfLongestSubstring/Q3_lengthOfLongestSubString.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
		vector<int> map(500, -1);

		int i = 0, j = 0, max = 0, longest = 0, flag=-1, num;
		
		while(i<s.length()){

			num = s[i];

			if(map[num] >= 0){
				i = map[num] + 1;
				fill(map.begin(), map.end(), -1);
				j = i;
				longest = 0;
				//cout << true;
			}
			else{
				map[num] = i;
				i++;
				longest++;
				if(longest > max){
					max = longest;
				}
				//cout << false;
			}
		}
		return max;
	}
};

File: data/leetCode Solutions/Q3_lengthOfLongestSubstring/Q3_lengthOfLongestSubString.js
// > 1st Way: BigO = n^3

// var lengthOfLongestSubstring = function(s) {
// 	let max = 0;
// 	if (s.length == 1)
// 		return 1;
	
// 	for (let i = 0; i < s.length; i++){
// 		const arr = [];
// 		let longest = 0;
// 		for (let j = i; j < s.length; j++){
// 			console.log(s[j]);
			
// 			if (arr.includes(s[j])) {
// 				break;
// 			}
// 			arr.push(s[j]);
// 			longest++;
// 		}
// 		if (max < longest) {
// 			max = longest;
// 		}
// 	}

// 	return max;
// };

// console.log(lengthOfLongestSubstring("c"));


// > 2nd Way

var lengthOfLongestSubstring = function (s)
{
	// s = Array.from(s);

	let max = 0, i = 0, flag, longest = 0;
	let arr = [];
	if (s.length == 1)
		return 1;
		
	for (let j = i; j < s.length; j++) {

		if (arr.includes(s[j])) {

			for (let k = i; k < j; k++){
				if (s[k] == s[j]) {
					flag = k;
				}
			}
			console.log('true');
				
			console.log(flag);
			arr = [];
			i = flag;
			j = i;
			longest = 0;
		}
		
		else {
			console.log('false');
			arr.push(s[j]);
			longest++;
			if (longest > max) {
				max = longest;
			}
		}
	}
	
	return max;
};
	
console.log(lengthOfLongestSubstring("bbbbb"));